File: packages/core/src/policy/approval-manifest.ts
/**
 * Approval Manifest for Matimo.
 *
 * Manages HMAC-signed approval records for agent-created tools.
 * Prevents agent forgery of approvals via cryptographic verification.
 */

import { createHmac, randomUUID, createHash } from 'crypto';
import fs from 'fs';
import path from 'path';
import { getGlobalMatimoLogger } from '../logging';

export interface ApprovalRecord {
  name: string;
  hash: string;
  signature: string;
  approvedAt: string;
  approvedBy?: string;
}

interface ManifestData {
  version: '1';
  approvals: ApprovalRecord[];
  pending?: string[];
}

export class ApprovalManifest {
  private readonly secret: string;
  private readonly manifestPath: string;
  private cache: Map<string, ApprovalRecord> = new Map();
  private pendingSet: Set<string> = new Set();

  /**
   * @param approvalDir - Directory where `.matimo-approvals.json` lives
   * @param secret - HMAC secret. If not provided, reads `MATIMO_APPROVAL_SECRET`
   *   from env. If that's also missing, generates one and logs it.
   */
  constructor(approvalDir: string, secret?: string) {
    this.manifestPath = path.join(approvalDir, '.matimo-approvals.json');

    if (secret) {
      this.secret = secret;
    } else if (process.env.MATIMO_APPROVAL_SECRET) {
      this.secret = process.env.MATIMO_APPROVAL_SECRET;
    } else {
      this.secret = randomUUID();
      const logger = getGlobalMatimoLogger();
      // Create a non-sensitive fingerprint for debugging (first 4 chars only)
      const fingerprint = this.secret.substring(0, 4);
      logger.warn(
        'No MATIMO_APPROVAL_SECRET set. An ephemeral secret was auto-generated ' +
          '(fingerprint: ' +
          fingerprint +
          '...) ' +
          'for this process. Approvals may not persist across restarts. ' +
          'To persist approvals, set MATIMO_APPROVAL_SECRET to a stable, securely ' +
          'generated value in the environment.'
      );
    }

    this.loadFromDisk();
  }

  /**
   * Compute the HMAC signature for a tool approval.
   */
  private sign(toolName: string, yamlHash: string): string {
    return createHmac('sha256', this.secret).update(`${toolName}:${yamlHash}`).digest('hex');
  }

  /**
   * Compute SHA-256 hash of content.
   */
  computeHash(content: string): string {
    return createHash('sha256').update(content, 'utf8').digest('hex');
  }

  /**
   * Verify that an approval record has a valid HMAC signature and
   * the stored hash matches the current YAML content hash.
   */
  isApproved(toolName: string, currentYamlHash: string): boolean {
    const record = this.cache.get(toolName);
    if (!record) return false;

    // Hash mismatch = YAML was modified after approval → revoked
    if (record.hash !== currentYamlHash) return false;

    // Verify HMAC to detect manifest tampering
    const expectedSig = this.sign(toolName, record.hash);
    return record.signature === expectedSig;
  }

  /**
   * Approve a tool. Creates an HMAC-signed record in the manifest.
   */
  approve(toolName: string, yamlHash: string, approvedBy?: string): void {
    const signature = this.sign(toolName, yamlHash);
    const record: ApprovalRecord = {
      name: toolName,
      hash: yamlHash,
      signature,
      approvedAt: new Date().toISOString(),
      approvedBy,
    };
    this.cache.set(toolName, record);
    this.saveToDisk();
  }

  /**
   * Revoke a tool's approval. Removes from both cache and pendingSet
   * to ensure consistent state (tool no longer tracked as approved or pending).
   */
  revoke(toolName: string): boolean {
    const cacheDeleted = this.cache.delete(toolName);
    const pendingDeleted = this.pendingSet.delete(toolName);
    if (cacheDeleted || pendingDeleted) {
      this.saveToDisk();
    }
    return cacheDeleted || pendingDeleted;
  }

  /**
   * Get the approval record for a tool.
   */
  getApproval(toolName: string): ApprovalRecord | undefined {
    return this.cache.get(toolName);
  }

  /**
   * List all approved tool names.
   */
  listApproved(): string[] {
    return Array.from(this.cache.keys());
  }

  /**
   * Mark a tool as pending approval. Called by matimo_create_tool after writing to disk.
   */
  markPending(toolName: string): void {
    this.pendingSet.add(toolName);
    this.saveToDisk();
  }

  /**
   * Return all tool names that have been proposed (written to disk) but not yet approved.
   */
  getPendingTools(): string[] {
    return Array.from(this.pendingSet).filter((name) => !this.cache.has(name));
  }

  /**
   * Load the manifest file from disk.
   */
  private loadFromDisk(): void {
    try {
      if (!fs.existsSync(this.manifestPath)) return;
      const raw = fs.readFileSync(this.manifestPath, 'utf-8');
      const data: ManifestData = JSON.parse(raw);
      if (data.version !== '1' || !Array.isArray(data.approvals)) return;

      this.cache.clear();
      for (const record of data.approvals) {
        if (record.name && record.hash && record.signature) {
          this.cache.set(record.name, record);
        }
      }

      this.pendingSet.clear();
      if (Array.isArray(data.pending)) {
        for (const name of data.pending) {
          if (typeof name === 'string') this.pendingSet.add(name);
        }
      }
    } catch {
      // Corrupted manifest — start fresh
      const logger = getGlobalMatimoLogger();
      logger.warn('Failed to load approval manifest, starting fresh', {
        path: this.manifestPath,
      });
      this.cache.clear();
      this.pendingSet.clear();
    }
  }

  /**
   * Save current approvals to disk using atomic write pattern.
   * Writes to a temporary file first, then atomically renames it.
   * This prevents data corruption if the process crashes mid-write.
   */
  private saveToDisk(): void {
    const data: ManifestData = {
      version: '1',
      approvals: Array.from(this.cache.values()),
      pending: Array.from(this.pendingSet).filter((name) => !this.cache.has(name)),
    };
    const dir = path.dirname(this.manifestPath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }

    // Atomic write: write to temp file, then rename
    const json = JSON.stringify(data, null, 2);
    const tempPath = path.join(
      dir,
      `${path.basename(this.manifestPath)}.tmp-${process.pid}-${Date.now()}`
    );
    fs.writeFileSync(tempPath, json, 'utf-8');
    fs.renameSync(tempPath, this.manifestPath);
  }
}

File: python/packages/core/src/matimo/policy/approval_manifest.py
"""
Approval manifest — HMAC-signed on-disk record of approved tools.
Mirrors: packages/core/src/policy/approval-manifest.ts
"""
from __future__ import annotations

import hashlib
import hmac
import json
import logging
import os
import tempfile
import uuid
from dataclasses import asdict, dataclass
from datetime import UTC
from pathlib import Path

logger = logging.getLogger("matimo")

_MANIFEST_FILENAME = ".matimo-approvals.json"


@dataclass
class ApprovalRecord:
    name: str
    hash: str
    signature: str   # HMAC-SHA256 hex of "{name}:{hash}"
    approved_at: str
    approved_by: str | None = None


class ApprovalManifest:
    """
    Persists HMAC-signed tool approval records to disk.
    Mirrors: ApprovalManifest in approval-manifest.ts

    Uses an ephemeral UUID signing key if MATIMO_APPROVAL_SECRET is not set.
    Ephemeral keys mean approvals do not survive process restarts — which is
    the safe default.

    Security:
    - Only the first 4 chars of the signing key are logged (fingerprint only).
    - Atomic write prevents manifest corruption on crash.
    """

    def __init__(
        self,
        approval_dir: str | None = None,
        approval_secret: str | None = None,
    ) -> None:
        self._dir = Path(
            approval_dir or os.environ.get("MATIMO_APPROVAL_DIR", ".")
        )
        raw_secret = (
            approval_secret
            or os.environ.get("MATIMO_APPROVAL_SECRET")
            or self._generate_ephemeral_secret()
        )
        self._secret = raw_secret.encode("utf-8")
        fingerprint = raw_secret[:4] + "****"
        logger.debug("ApprovalManifest initialised (key fingerprint: %s)", fingerprint)

        self._manifest_path = self._dir / _MANIFEST_FILENAME
        self._records: dict[str, ApprovalRecord] = {}
        self._load()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def is_approved(self, tool_name: str, content_hash: str) -> bool:
        """Return True if tool_name has a valid approval for content_hash."""
        record = self._records.get(tool_name)
        if record is None:
            return False
        if record.hash != content_hash:
            return False
        return self._verify_signature(record)

    def approve(
        self,
        tool_name: str,
        content_hash: str,
        approved_by: str | None = None,
    ) -> ApprovalRecord:
        """Record a new approval for a tool."""
        from datetime import datetime

        signature = self._sign(tool_name, content_hash)
        record = ApprovalRecord(
            name=tool_name,
            hash=content_hash,
            signature=signature,
            approved_at=datetime.now(UTC).isoformat(),
            approved_by=approved_by,
        )
        self._records[tool_name] = record
        self._save()
        logger.info("Tool '%s' approved (hash: %s...)", tool_name, content_hash[:8])
        return record

    def revoke(self, tool_name: str) -> bool:
        """Remove the approval for a tool. Returns True if it existed."""
        if tool_name not in self._records:
            return False
        del self._records[tool_name]
        self._save()
        logger.info("Approval revoked for tool '%s'", tool_name)
        return True

    def get_all(self) -> list[ApprovalRecord]:
        return list(self._records.values())

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _sign(self, tool_name: str, content_hash: str) -> str:
        message = f"{tool_name}:{content_hash}".encode()
        return hmac.new(self._secret, message, hashlib.sha256).hexdigest()

    def _verify_signature(self, record: ApprovalRecord) -> bool:
        expected = self._sign(record.name, record.hash)
        return hmac.compare_digest(expected, record.signature)

    def _load(self) -> None:
        if not self._manifest_path.exists():
            return
        try:
            raw = self._manifest_path.read_text(encoding="utf-8")
            data = json.loads(raw)
            for entry in data.get("approvals", []):
                record = ApprovalRecord(**entry)
                if self._verify_signature(record):
                    self._records[record.name] = record
                else:
                    logger.warning(
                        "Approval record for '%s' has invalid signature — discarding",
                        entry.get("name", "?"),
                    )
        except Exception as exc:
            logger.warning("Failed to load approval manifest: %s", exc)

    def _save(self) -> None:
        """Atomic write to prevent manifest corruption on crash."""
        self._dir.mkdir(parents=True, exist_ok=True)
        payload = {
            "version": "1",
            "approvals": [asdict(r) for r in self._records.values()],
        }
        tmp_fd, tmp_path = tempfile.mkstemp(
            dir=self._dir, prefix=".matimo-approvals-tmp-"
        )
        try:
            with os.fdopen(tmp_fd, "w", encoding="utf-8") as f:
                json.dump(payload, f, indent=2)
            os.replace(tmp_path, self._manifest_path)
        except Exception as exc:
            # Clean up temp file on failure
            try:
                os.unlink(tmp_path)
            except OSError:
                pass
            logger.error("Failed to save approval manifest: %s", exc)

    @staticmethod
    def _generate_ephemeral_secret() -> str:
        """Generate a random ephemeral signing key (not persistent across restarts)."""
        return str(uuid.uuid4())
